File: human-solutions/05/python/main.py
from pathlib import Path

input_text = Path("../../../inputs/05.txt").read_text().strip()

sections = input_text.split("\n\n")

ranges = [
    list(map(int, line.split("-")))
    for line in sections[0].splitlines()
]

ids = list(map(int, sections[1].splitlines()))

fresh_ingredients = [
    id_
    for id_ in ids
    if any(min_ <= id_ <= max_ for min_, max_ in ranges)
]

print(len(fresh_ingredients))

ranges = [(min(a, b), max(a, b)) for a, b in ranges]
ranges.sort(key=lambda r: r[0])

merged = []

for start, end in ranges:
    if not merged or start > merged[-1][1] + 1:
        merged.append([start, end])
    else:
        merged[-1][1] = max(merged[-1][1], end)

total_fresh_ids = sum(end - start + 1 for start, end in merged)

print(total_fresh_ids)

File: human-solutions/05/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("../../../inputs/05.txt")
        .expect("read error")
        .trim()
        .to_string();

    let sections: Vec<&str> = input.split("\n\n").collect();

    let mut ranges: Vec<(i64, i64)> = sections[0]
        .lines()
        .map(|line| {
            let mut it = line.split('-').map(|n| n.parse::<i64>().unwrap());
            let a = it.next().unwrap();
            let b = it.next().unwrap();
            (a.min(b), a.max(b))
        })
        .collect();

    let ids: Vec<i64> = sections[1]
        .lines()
        .map(|line| line.parse::<i64>().unwrap())
        .collect();

    let fresh_ingredients: Vec<i64> = ids
        .iter()
        .copied()
        .filter(|id| ranges.iter().any(|&(min, max)| *id >= min && *id <= max))
        .collect();

    println!("{}", fresh_ingredients.len());

    ranges.sort_by_key(|(start, _)| *start);

    let mut merged: Vec<(i64, i64)> = Vec::new();

    for (start, end) in ranges {
        match merged.last_mut() {
            None => merged.push((start, end)),
            Some((_, last_end)) => {
                if start > *last_end + 1 {
                    merged.push((start, end));
                } else {
                    *last_end = (*last_end).max(end);
                }
            }
        }
    }

    let total_fresh_ids: i64 = merged
        .iter()
        .map(|(start, end)| end - start + 1)
        .sum();

    println!("{}", total_fresh_ids);
}
